### DISCUSSION TITLE
How do clients support robust CSS3 and still ensure security for their users?

### DISCUSSION THREAD

**Discussion by hteumeuleu** [email-markup-consortium/email-markup-consortium#17] [General]
A conversation recently started on the [#emailgeeks Slack](https://email.geeks.chat/) by @alicemeichi:

> There were a few points brought up that the reason why we can’t just support web-standard markup in email is due to security reasons and concerns around protecting users, which is totally valid.However, how do iOS/MacOS clients like Apple/iPhone/iPad Mail support robust CSS3 and still ensure security for their users? Not to mention, Apple now features one of the most user-focused privacy policies out there now (leading to the recent Marketing meltdown over MPP). Any insights on this would be awesome
>
> What can an app do to ensure security that webmail can’t? Gmail has a mobile app, but it’s arguably even more restrictive than Gmail webmail in terms of the markup it supports.

To which I replied: 

> This is the biggest key here. Apps versus webmails. Apps still need to filter and remove dangerous things (like JavaScript, `<meta>` tags and such), but they can be way more tolerant than webmails. For example, there’s no problem for an app to support `position:absolute` while it can be a serious security issue for a webmail. (An attacker could use this to position elements from an email over the whole webmail interface.)
>
> Another thing to consider is the way email clients would filter or block such harmful HTML and CSS features. There are two ways to do so: either using a block list (so you're supporting everything by default but blocking a few features) or a safe list (so you're blocking everything by default but allowing a few features). Due to the ever evolving nature of the Web, I think that webmail clients that use a block list are at a serious risk and they all should use a safe list instead. For example, there's an unlimited and ever growing number of JavaScript events available as HTML attributes (like `onmouseover`, `onload`, …). But new events attributes like this pop up all the time, sometimes only on specific browsers, or other times removed in newer browsers but still present in older browsers. If you're a webmail developer, this means that either you can commit to maintain every day an ever growing block list or have a safe list that you update every now and then.
>
> As for Gmail, I think this falls in the way they chose to filter HTML emails server side and have common filter across their clients. It's easier for them to do it this way, but this means they have less support overall in their apps than what they could potentially have.

@alicemeichi replied:

> For example, there’s no problem for an app to support `position:absolute` while it can be a serious security issue for a webmail. (An attacker could use this to position elements from an email over the whole webmail interface.)
> 
> I don’t understand this either, @hteumeuleu. I’m working on a team that develops a single page web app ESP, and we implement all sorts of security measures to prevent XSS attacks and the like. Being that Gmail is also a single page web app, I still don’t understand why they can’t do the same. I think @jkupczak probably hit it on the nose when it comes to the “indifference” factor.Okay here’s an example of another webmail client (not an app) that features Apple Mail-level of markup support — Hey.com. How do they do it?

I replied:

> I think the heart of the problem is that most of the security suggestions here are applied client-side. Which means that any third-party (like a malicious browser extension) could override this easily. I don't know what motivates Gmail to not improve their support for CSS, and I'm only guessing as an email developer, but I'm sure the fact that they serve over 1.5 billion active users and a lot of governmental agencies around the world make them subject to a lot of very specific and custom attacks against their users. HEY's desktop webmail can support more stuff because they don't have to support IE6 and they probably don't have any governmental customers.By the way, I'm loving this discussion because it feels like we're getting at the heart of things for standards recommendations. But shouldn't we talk about this more publicly (on the group's GitHub for example) so that, hopefully, email clients developers can chime in as well and enlighten us?

And here we are! What are your thoughts/knowledge on the subject?

**Comment by jk** [comment]
@hteumeuleu I'm most probably not the one you want to mention. Do not carbon copy mails into github.

**Reply by hteumeuleu** [reply]
Oops! Indeed, I just copied from a Slack group. I fixed it above. Thanks.
